src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -370,126 +418,410 @@ private sealed class SendFileOperation : WriteOperation
-
-                if (!IsEmpty)
+                using (Lock())
                 {
-                    operation.Next = _tail.Next;
-                    _tail.Next = operation;
-                }
+                    observedSequenceNumber = _sequenceNumber;
+                    bool isReady = (_state == QueueState.Ready);
 
-                _tail = operation;
+                    if (TraceEnabled) Trace(context, $"{isReady}");
+
+                    return isReady;
+                }
             }
 
-            private bool TryDequeue(out TOperation operation)
+            // Return true for pending, false for completed synchronously (including failure and abort)
+            public bool StartAsyncOperation(SocketAsyncContext context, TOperation operation, int observedSequenceNumber)
             {
-                if (_tail == null)
+                if (TraceEnabled) Trace(context, $"Enter");
+
+                if (!context._registered)
                 {
-                    operation = null;
-                    return false;
+                    context.Register();
                 }
 
-                AsyncOperation head = _tail.Next;
-                if (head == _tail)
+                while (true)
                 {
-                    _tail = null;
+                    bool doAbort = false;
+                    using (Lock())
+                    {
+                        switch (_state)
+                        {
+                            case QueueState.Ready:
+                                if (observedSequenceNumber != _sequenceNumber)
+                                {
+                                    // The queue has become ready again since we previously checked it.
+                                    // So, we need to retry the operation before we enqueue it.
+                                    Debug.Assert(observedSequenceNumber - _sequenceNumber < 10000, "Very large sequence number increase???");
+                                    observedSequenceNumber = _sequenceNumber;
+                                    break;
+                                }
+
+                                // Caller tried the operation and got an EWOULDBLOCK, so we need to transition.
+                                _state = QueueState.Waiting;
+                                goto case QueueState.Waiting;
+
+                            case QueueState.Waiting:
+                            case QueueState.Processing:
+                                // Enqueue the operation.
+                                Debug.Assert(operation.Next == operation, "Expected operation.Next == operation");
+
+                                if (_tail != null)
+                                {
+                                    operation.Next = _tail.Next;
+                                    _tail.Next = operation;
+                                }
+
+                                _tail = operation;
```

Comment:
It's a circular list, and on Operation creation, operation.Next is initialized to operation.  There's an assert about this above, though it may not be clear why the assert is true.

I'm not sure it does us any good to have a circular list here, but (a) that's the way the previous code worked, and (b) it does save us a ref field on the queue struct.  At one point I had changed this to not be circular, but that got lost along the way somewhere.